datafusion/core/src/logical_plan/builder.rs

```diff
@@ -432,8 +432,50 @@ impl LogicalPlanBuilder {
     }
 
     /// Apply a union, preserving duplicate rows
-    pub fn union(&self, plan: LogicalPlan) -> Result<Self> {
-        Ok(Self::from(union_with_alias(self.plan.clone(), plan, None)?))
+    pub fn union(&self, right_plan: LogicalPlan) -> Result<Self> {
+        let left_plan = self.plan.clone();
+        let union_schema = left_plan.schema().clone();
+        let inputs_iter = vec![left_plan, right_plan]
+            .into_iter()
+            .flat_map(|p| match p {
+                LogicalPlan::Union(Union { inputs, .. }) => inputs,
+                x => vec![x],
+            });
+
+        inputs_iter
+            .clone()
+            .skip(1)
+            .try_for_each(|input_plan| -> Result<()> {
+                union_schema.check_arrow_schema_type_compatible(
+                    &((**input_plan.schema()).clone().into()),
```

Comment:
`&((**` is definitely hard on the eyes (though I realize it wasn't introduced in this PR)